controllers/managedclusterForCGU_controller.go

```diff
@@ -0,0 +1,248 @@
+//
+// Note: The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ManagedClusterForCguReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	reqLogger := r.Log.WithValues("Request.Name", req.Name)
+	reqLogger.Info("Reconciling managedCluster to create clusterGroupUpgrade")
+
+	managedCluster := &clusterv1.ManagedCluster{}
+	if err := r.Get(ctx, types.NamespacedName{Name: req.Name}, managedCluster); err != nil {
+		if errors.IsNotFound(err) {
+			// managed cluster could have been deleted, delete its clusterCroupUpgrade
+			r.Log.Info("cluster not found", "name", req.Name)
+			clusterGroupUpgrade := &ranv1alpha1.ClusterGroupUpgrade{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      req.Name,
+					Namespace: ztpInstallNS,
+				},
+			}
+
+			r.Log.Info("Deleting clusterGroupUpgrade", "name", req.Name, "namespace", ztpInstallNS)
+			if err := r.Delete(ctx, clusterGroupUpgrade); err != nil {
+				if errors.IsNotFound(err) {
+					// clusterGroupUprade for the cluster doesn't exist/could have been deleted
+					return ctrl.Result{}, nil
+				}
+				// Error deleting clusterGroupUpgrade, requeue the request
+				r.Log.Error(err, "Fail to delete clusterGroupUpgrade", "name", req.Name, "namespace", ztpInstallNS)
+				return ctrl.Result{}, err
+			}
+			// clusterGroupUpgrade deleted, stop reconcile
+			return ctrl.Result{}, nil
+		}
+		// Error reading managed cluster, requeue the request
+		return ctrl.Result{}, err
+	}
+
+	clusterGroupUpgrade := &ranv1alpha1.ClusterGroupUpgrade{}
+	if err := r.Get(ctx, types.NamespacedName{Name: req.Name, Namespace: ztpInstallNS}, clusterGroupUpgrade); err != nil {
+		if !errors.IsNotFound(err) {
+			// Error reading clusterGroupUpgrade, requeue the request
+			return ctrl.Result{}, err
+		}
+	} else {
+		// clusterGroupUpgrade for this cluster already exists, stop reconcile
+		r.Log.Info("clusterGroupUpgrade found", "Name", clusterGroupUpgrade.Name, "Namespace", clusterGroupUpgrade.Namespace)
+		return ctrl.Result{}, nil
+	}
+
+	// clusterGroupUpgrade CR doesn't exist
+	availableCondition := meta.FindStatusCondition(managedCluster.Status.Conditions, clusterv1.ManagedClusterConditionAvailable)
+	if availableCondition == nil {
+		r.Log.Info("cluster has no status yet", "Name", managedCluster.Name, "RequeueAfter:", clusterStatusCheckRetryDelay)
+		return ctrl.Result{RequeueAfter: clusterStatusCheckRetryDelay}, nil
+	} else if availableCondition.Status == metav1.ConditionTrue {
+		// cluster is ready
+
+		// get child policies for the cluster
+		// child polices created as soon as placementrule returns the cluster in decision
+		// and placementrule returns decision when cluster is ready
+		policies, err := r.getPolicies(ctx, managedCluster.Name)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if len(policies.Items) == 0 {
```

Comment:
This may create a corner-case where a user-created ManagedCluster which has no policies associated with it would be requeued forever. For now I'd say just log this as a warning and don't requeue.

Reply:
Ok. Will update